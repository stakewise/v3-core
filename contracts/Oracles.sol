// SPDX-License-Identifier: BUSL-1.1

pragma solidity =0.8.17;

import {Ownable} from '@openzeppelin/contracts/access/Ownable.sol';
import {EIP712} from '@openzeppelin/contracts/utils/cryptography/EIP712.sol';
import {ECDSA} from '@openzeppelin/contracts/utils/cryptography/ECDSA.sol';
import {IOracles} from './interfaces/IOracles.sol';

/**
 * @title Oracles
 * @author StakeWise
 * @notice Defines the functionality for verifying signatures of the whitelisted off-chain oracles
 */
contract Oracles is Ownable, EIP712, IOracles {
  /// @inheritdoc IOracles
  mapping(address => bool) public override isOracle;

  /// @inheritdoc IOracles
  uint256 public override totalOracles;

  /// @inheritdoc IOracles
  uint256 public override requiredOracles;

  /**
   * @dev Constructor
   * @param _owner The address of the contract owner
   * @param initialOracles The addresses of the initial oracles
   * @param initialRequiredOracles The number or required oracles for the verification
   */
  constructor(
    address _owner,
    address[] memory initialOracles,
    uint256 initialRequiredOracles
  ) EIP712('Oracles', '1') {
    for (uint256 i = 0; i < initialOracles.length; ) {
      addOracle(initialOracles[i]);
      unchecked {
        ++i;
      }
    }
    setRequiredOracles(initialRequiredOracles);
    _transferOwnership(_owner);
  }

  /// @inheritdoc IOracles
  function addOracle(address oracle) public override onlyOwner {
    if (isOracle[oracle]) revert AlreadyAdded();

    isOracle[oracle] = true;
    unchecked {
      // cannot realistically overflow
      totalOracles += 1;
    }
    emit OracleAdded(oracle);
  }

  /// @inheritdoc IOracles
  function removeOracle(address oracle) external override onlyOwner {
    if (!isOracle[oracle]) revert AlreadyRemoved();

    // SLOAD to memory
    uint256 _totalOracles;
    unchecked {
      // cannot underflow
      _totalOracles = totalOracles - 1;
    }

    isOracle[oracle] = false;
    totalOracles = _totalOracles;
    if (_totalOracles < requiredOracles) setRequiredOracles(_totalOracles);

    emit OracleRemoved(oracle);
  }

  /// @inheritdoc IOracles
  function setRequiredOracles(uint256 _requiredOracles) public override onlyOwner {
    if (_requiredOracles == 0 || totalOracles < _requiredOracles) revert InvalidRequiredOracles();
    requiredOracles = _requiredOracles;
    emit RequiredOraclesUpdated(_requiredOracles);
  }

  /// @inheritdoc IOracles
  function verifySignatures(bytes32 message, bytes calldata signatures) external view override {
    // SLOAD to memory
    uint256 _requiredOracles = requiredOracles;

    // check whether enough signatures
    if (signatures.length < _requiredOracles * 65) revert NotEnoughSignatures();

    bytes32 data = _hashTypedDataV4(message);
    address lastOracle;
    address currentOracle;
    uint256 startIndex;
    for (uint256 i = 0; i < _requiredOracles; ) {
      unchecked {
        // cannot overflow as signatures.length is checked above
        currentOracle = ECDSA.recover(data, signatures[startIndex:startIndex + 65]);
      }
      if (currentOracle <= lastOracle || !isOracle[currentOracle]) revert InvalidOracle();

      // update last oracle
      lastOracle = currentOracle;

      unchecked {
        // cannot overflow as it's capped with requiredOracles
        ++i;
        startIndex += 65;
      }
    }
  }
}
