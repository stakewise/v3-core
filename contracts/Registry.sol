// SPDX-License-Identifier: BUSL-1.1

pragma solidity =0.8.17;

import {Ownable} from '@openzeppelin/contracts/access/Ownable.sol';
import {IRegistry} from './interfaces/IRegistry.sol';

/**
 * @title Registry
 * @author StakeWise
 * @notice Defines the registry functionality that keeps track of vaults, factories and their upgrades
 */
contract Registry is Ownable, IRegistry {
  /// @inheritdoc IRegistry
  mapping(address => bool) public override vaults;

  /// @inheritdoc IRegistry
  mapping(address => bool) public override factories;

  /// @inheritdoc IRegistry
  mapping(address => address) public override upgrades;

  /**
   * @dev Constructor
   * @param owner_ The address of the registry owner
   */
  constructor(address owner_) {
    _transferOwnership(owner_);
  }

  /// @inheritdoc IRegistry
  function addVault(address vault) external override {
    if (!factories[msg.sender]) revert AccessDenied();
    vaults[vault] = true;
    emit VaultAdded(msg.sender, vault);
  }

  /// @inheritdoc IRegistry
  function addUpgrade(address prevImpl, address newImpl) external override onlyOwner {
    if (prevImpl == newImpl) revert InvalidUpgrade();
    upgrades[prevImpl] = newImpl;
    emit UpgradeAdded(prevImpl, newImpl);
  }

  /// @inheritdoc IRegistry
  function addFactory(address factory) external override onlyOwner {
    if (factories[factory]) revert AlreadyAdded();
    factories[factory] = true;
    emit FactoryAdded(factory);
  }

  /// @inheritdoc IRegistry
  function removeFactory(address factory) external override onlyOwner {
    if (!factories[factory]) revert AlreadyRemoved();
    factories[factory] = false;
    emit FactoryRemoved(factory);
  }
}
