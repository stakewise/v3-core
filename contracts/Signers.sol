// SPDX-License-Identifier: BUSL-1.1

pragma solidity =0.8.17;

import {Ownable} from '@openzeppelin/contracts/access/Ownable.sol';
import {EIP712} from '@openzeppelin/contracts/utils/cryptography/EIP712.sol';
import {ECDSA} from '@openzeppelin/contracts/utils/cryptography/ECDSA.sol';
import {ISigners} from './interfaces/ISigners.sol';

/**
 * @title Signers
 * @author StakeWise
 * @notice Defines the functionality for verifying signatures of the whitelisted off-chain signers
 */
contract Signers is Ownable, EIP712, ISigners {
  /// @inheritdoc ISigners
  mapping(address => bool) public override isSigner;

  /// @inheritdoc ISigners
  uint256 public override totalSigners;

  /// @inheritdoc ISigners
  uint256 public override requiredSigners;

  /**
   * @dev Constructor
   * @param _owner The address of the contract owner
   * @param initialSigners The addresses of the initial signers
   * @param initialRequiredSigners The number or required signers for the verification
   */
  constructor(
    address _owner,
    address[] memory initialSigners,
    uint256 initialRequiredSigners
  ) EIP712('Signers', '1') {
    for (uint256 i = 0; i < initialSigners.length; ) {
      addSigner(initialSigners[i]);
      unchecked {
        ++i;
      }
    }
    setRequiredSigners(initialRequiredSigners);
    _transferOwnership(_owner);
  }

  /// @inheritdoc ISigners
  function addSigner(address signer) public override onlyOwner {
    if (isSigner[signer]) revert AlreadyAdded();

    isSigner[signer] = true;
    unchecked {
      // cannot realistically overflow
      totalSigners += 1;
    }
    emit SignerAdded(signer);
  }

  /// @inheritdoc ISigners
  function removeSigner(address signer) external override onlyOwner {
    if (!isSigner[signer]) revert AlreadyRemoved();

    // SLOAD to memory
    uint256 _totalSigners;
    unchecked {
      // cannot underflow
      _totalSigners = totalSigners - 1;
    }

    isSigner[signer] = false;
    totalSigners = _totalSigners;
    if (_totalSigners < requiredSigners) setRequiredSigners(_totalSigners);

    emit SignerRemoved(signer);
  }

  /// @inheritdoc ISigners
  function setRequiredSigners(uint256 _requiredSigners) public override onlyOwner {
    if (_requiredSigners == 0 || totalSigners < _requiredSigners) revert InvalidRequiredSigners();
    requiredSigners = _requiredSigners;
    emit RequiredSignersUpdated(_requiredSigners);
  }

  /// @inheritdoc ISigners
  function verifySignatures(bytes32 message, bytes calldata signatures) external view override {
    // SLOAD to memory
    uint256 _requiredSigners = requiredSigners;

    // check whether enough signatures
    if (signatures.length < _requiredSigners * 65) revert NotEnoughSignatures();

    bytes32 data = _hashTypedDataV4(message);
    address lastSigner;
    address currentSigner;
    uint256 startIndex;
    for (uint256 i = 0; i < _requiredSigners; ) {
      unchecked {
        // cannot overflow as signatures.length is checked above
        currentSigner = ECDSA.recover(data, signatures[startIndex:startIndex + 65]);
      }
      if (currentSigner <= lastSigner || !isSigner[currentSigner]) revert InvalidSigner();

      // update last signer
      lastSigner = currentSigner;

      unchecked {
        // cannot overflow as it's capped with requiredSigners
        ++i;
        startIndex += 65;
      }
    }
  }
}
