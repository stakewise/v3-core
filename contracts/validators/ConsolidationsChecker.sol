// SPDX-License-Identifier: BUSL-1.1

pragma solidity ^0.8.22;

import {ECDSA} from '@openzeppelin/contracts/utils/cryptography/ECDSA.sol';
import {EIP712} from '@openzeppelin/contracts/utils/cryptography/EIP712.sol';
import {IConsolidationsChecker} from '../interfaces/IConsolidationsChecker.sol';
import {Errors} from '../libraries/Errors.sol';
import {IKeeper} from '../interfaces/IKeeper.sol';

/**
 * @title ConsolidationsChecker
 * @author StakeWise
 * @notice Defines the functionality for checking signatures of oracles for validators consolidations
 */
contract ConsolidationsChecker is EIP712, IConsolidationsChecker {
  uint256 private constant _signatureLength = 65;
  bytes32 private constant _consolidationsCheckerTypeHash =
    keccak256('ConsolidationsChecker(address vault,bytes validators)');

  IKeeper private immutable _keeper;

  /**
   * @dev Constructor
   * @param keeper The address of the Keeper contract
   */
  constructor(address keeper) EIP712('ConsolidationsChecker', '1') {
    _keeper = IKeeper(keeper);
  }

  /// @inheritdoc IConsolidationsChecker
  function verifySignatures(
    address vault,
    bytes calldata validators,
    bytes calldata signatures
  ) external view override {
    if (!isValidSignatures(vault, validators, signatures)) {
      revert Errors.InvalidSignatures();
    }
  }

  /// @inheritdoc IConsolidationsChecker
  function isValidSignatures(
    address vault,
    bytes calldata validators,
    bytes calldata signatures
  ) public view override returns (bool) {
    return
      _isValidSignatures(
        _keeper.validatorsMinOracles(),
        keccak256(abi.encode(_consolidationsCheckerTypeHash, vault, keccak256(validators))),
        signatures
      );
  }

  /**
   * @notice Internal function for verifying oracles' signatures
   * @param requiredSignatures The number of signatures required for the verification to pass
   * @param message The message that was signed
   * @param signatures The concatenation of the oracles' signatures
   * @return True if the signatures are valid, otherwise false
   */
  function _isValidSignatures(
    uint256 requiredSignatures,
    bytes32 message,
    bytes calldata signatures
  ) private view returns (bool) {
    if (requiredSignatures == 0) {
      return false;
    }

    // check whether enough signatures
    unchecked {
      // cannot realistically overflow
      if (signatures.length < requiredSignatures * _signatureLength) {
        return false;
      }
    }

    bytes32 data = _hashTypedDataV4(message);
    address lastOracle;
    address currentOracle;
    uint256 startIndex;
    for (uint256 i = 0; i < requiredSignatures; i++) {
      unchecked {
        // cannot overflow as signatures.length is checked above
        currentOracle = ECDSA.recover(data, signatures[startIndex:startIndex + _signatureLength]);
      }
      // signatures must be sorted by oracles' addresses and not repeat
      if (currentOracle <= lastOracle || !_keeper.isOracle(currentOracle)) {
        return false;
      }

      // update last oracle
      lastOracle = currentOracle;

      unchecked {
        // cannot realistically overflow
        startIndex += _signatureLength;
      }
    }

    return true;
  }
}
