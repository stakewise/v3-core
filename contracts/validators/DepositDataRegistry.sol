// SPDX-License-Identifier: BUSL-1.1

pragma solidity =0.8.22;

import {MerkleProof} from '@openzeppelin/contracts/utils/cryptography/MerkleProof.sol';
import {IDepositDataRegistry} from '../interfaces/IDepositDataRegistry.sol';
import {IKeeperValidators} from '../interfaces/IKeeperValidators.sol';
import {IVaultAdmin} from '../interfaces/IVaultAdmin.sol';
import {IVaultValidators} from '../interfaces/IVaultValidators.sol';
import {IVaultVersion} from '../interfaces/IVaultVersion.sol';
import {IVaultsRegistry} from '../interfaces/IVaultsRegistry.sol';
import {IVaultState} from '../interfaces/IVaultState.sol';
import {IKeeperRewards} from '../interfaces/IKeeperRewards.sol';
import {Errors} from '../libraries/Errors.sol';
import {Multicall} from '../base/Multicall.sol';

/**
 * @title DepositDataRegistry
 * @author StakeWise
 * @notice Defines the functionality for the Vault's deposit data management
 */
contract DepositDataRegistry is Multicall, IDepositDataRegistry {
  IVaultsRegistry private immutable _vaultsRegistry;

  /// @inheritdoc IDepositDataRegistry
  mapping(address => uint256) public override depositDataIndexes;

  /// @inheritdoc IDepositDataRegistry
  mapping(address => bytes32) public override depositDataRoots;

  mapping(address => address) private _depositDataManagers;
  mapping(address => bool) private _migrated;

  /**
   * @dev Modifier to check if the caller is a valid vault
   * @param vault The address of the vault
   */
  modifier onlyValidVault(address vault) {
    if (!_vaultsRegistry.vaults(vault) || IVaultVersion(vault).version() < 2) {
      revert Errors.InvalidVault();
    }
    _;
  }

  /**
   * @dev Constructor
   * @param vaultsRegistry The address of the vaults registry contract
   */
  constructor(address vaultsRegistry) {
    _vaultsRegistry = IVaultsRegistry(vaultsRegistry);
  }

  /// @inheritdoc IDepositDataRegistry
  function getDepositDataManager(address vault) public view override returns (address) {
    address depositDataManager = _depositDataManagers[vault];
    return depositDataManager == address(0) ? IVaultAdmin(vault).admin() : depositDataManager;
  }

  /// @inheritdoc IDepositDataRegistry
  function setDepositDataManager(
    address vault,
    address depositDataManager
  ) external override onlyValidVault(vault) {
    // only vault admin can set deposit data manager
    if (msg.sender != IVaultAdmin(vault).admin()) revert Errors.AccessDenied();

    // update deposit data manager
    _depositDataManagers[vault] = depositDataManager;
    emit DepositDataManagerUpdated(vault, depositDataManager);
  }

  /// @inheritdoc IDepositDataRegistry
  function setDepositDataRoot(
    address vault,
    bytes32 depositDataRoot
  ) external override onlyValidVault(vault) {
    if (msg.sender != getDepositDataManager(vault)) revert Errors.AccessDenied();
    if (depositDataRoots[vault] == depositDataRoot) revert Errors.ValueNotChanged();

    depositDataRoots[vault] = depositDataRoot;
    // reset validator index on every root update
    depositDataIndexes[vault] = 0;
    emit DepositDataRootUpdated(vault, depositDataRoot);
  }

  /// @inheritdoc IDepositDataRegistry
  function updateVaultState(
    address vault,
    IKeeperRewards.HarvestParams calldata harvestParams
  ) external override {
    IVaultState(vault).updateState(harvestParams);
  }

  /// @inheritdoc IDepositDataRegistry
  function registerValidator(
    address vault,
    IKeeperValidators.ApprovalParams calldata keeperParams,
    bytes32[] calldata proof
  ) external override onlyValidVault(vault) {
    // register validator
    IVaultValidators(vault).registerValidators(keeperParams, '');

    // SLOAD to memory
    uint256 currentIndex = depositDataIndexes[vault];
    bytes32 depositDataRoot = depositDataRoots[vault];

    // check matches merkle root and next validator index
    if (
      !MerkleProof.verifyCalldata(
        proof,
        depositDataRoot,
        keccak256(bytes.concat(keccak256(abi.encode(keeperParams.validators, currentIndex))))
      )
    ) {
      revert Errors.InvalidProof();
    }

    // increment index for the next validator
    unchecked {
      // cannot realistically overflow
      depositDataIndexes[vault] = currentIndex + 1;
    }
  }

  /// @inheritdoc IDepositDataRegistry
  function registerValidators(
    address vault,
    IKeeperValidators.ApprovalParams calldata keeperParams,
    uint256[] calldata indexes,
    bool[] calldata proofFlags,
    bytes32[] calldata proof
  ) external override onlyValidVault(vault) {
    // register validator
    IVaultValidators(vault).registerValidators(keeperParams, '');

    // SLOAD to memory
    uint256 currentIndex = depositDataIndexes[vault];
    bytes32 depositDataRoot = depositDataRoots[vault];

    // define leaves for multiproof
    uint256 validatorsCount = indexes.length;
    if (validatorsCount == 0) revert Errors.InvalidValidators();
    bytes32[] memory leaves = new bytes32[](validatorsCount);

    // calculate validator length
    uint256 validatorLength = keeperParams.validators.length / validatorsCount;
    if (validatorLength == 0) revert Errors.InvalidValidators();

    // calculate leaves
    {
      uint256 startIndex;
      uint256 endIndex;
      for (uint256 i = 0; i < validatorsCount; ) {
        endIndex += validatorLength;
        leaves[indexes[i]] = keccak256(
          bytes.concat(
            keccak256(abi.encode(keeperParams.validators[startIndex:endIndex], currentIndex))
          )
        );

        startIndex = endIndex;
        unchecked {
          // cannot realistically overflow
          ++currentIndex;
          ++i;
        }
      }
    }

    // check matches merkle root and next validator index
    if (!MerkleProof.multiProofVerifyCalldata(proof, proofFlags, depositDataRoot, leaves)) {
      revert Errors.InvalidProof();
    }

    // increment index for the next validator
    depositDataIndexes[vault] = currentIndex;
  }

  /// @inheritdoc IDepositDataRegistry
  function migrate(
    bytes32 depositDataRoot,
    uint256 validatorIndex,
    address depositDataManager
  ) external override onlyValidVault(msg.sender) {
    if (_migrated[msg.sender]) revert Errors.AccessDenied();

    depositDataRoots[msg.sender] = depositDataRoot;
    depositDataIndexes[msg.sender] = validatorIndex;
    _depositDataManagers[msg.sender] = depositDataManager;

    // only allow migration once
    _migrated[msg.sender] = true;
    emit DepositDataMigrated(msg.sender, depositDataRoot, validatorIndex, depositDataManager);
  }
}
