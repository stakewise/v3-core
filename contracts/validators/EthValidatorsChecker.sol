// SPDX-License-Identifier: BUSL-1.1

pragma solidity =0.8.22;

import {ValidatorsChecker} from './ValidatorsChecker.sol';

/**
 * @title EthValidatorsChecker
 * @author StakeWise
 * @notice Defines functionality for checking validators registration on Ethereum
 */
contract EthValidatorsChecker is ValidatorsChecker {
  /**
   * @dev Constructor
   * @param validatorsRegistry The address of the beacon chain validators registry contract
   * @param keeper The address of the Keeper contract
   * @param vaultsRegistry The address of the VaultsRegistry contract
   * @param depositDataRegistry The address of the DepositDataRegistry contract
   */
  constructor(
    address validatorsRegistry,
    address keeper,
    address vaultsRegistry,
    address depositDataRegistry
  ) ValidatorsChecker(validatorsRegistry, keeper, vaultsRegistry, depositDataRegistry) {}

  /// @inheritdoc ValidatorsChecker
  function _depositAmount() internal pure override returns (uint256) {
    return 32 ether;
  }
}
