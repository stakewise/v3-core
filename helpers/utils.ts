import { Contract } from 'ethers'
import { HardhatRuntimeEnvironment } from 'hardhat/src/types/runtime'

export async function deployContract(tx: any): Promise<Contract> {
  const result = await tx
  await result.deployTransaction.wait()
  return result
}

async function delay(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export async function verify(
  hre: HardhatRuntimeEnvironment,
  address: string,
  constructorArgs: any,
  contractPath: string
) {
  if (!process.env.BLOCK_EXPLORER_KEY) {
    return
  }

  let count = 0
  const maxTries = 8
  // eslint-disable-next-line no-constant-condition
  while (true) {
    await delay(10000)
    try {
      console.log('Verifying contract at', address)
      await hre.run('verify:verify', {
        address,
        constructorArguments: constructorArgs,
        contract: contractPath,
      })
      break
    } catch (error) {
      if (String(error).includes('Already Verified')) {
        console.log(`Already verified contract at ${contractPath} at address ${address}`)
        break
      }
      if (++count == maxTries) {
        console.log(
          `Failed to verify contract at ${contractPath} at address ${address}, error: ${error}`
        )
        break
      }
      console.log(`Retrying... Retry #${count}, last error: ${error}`)
    }
  }
}
