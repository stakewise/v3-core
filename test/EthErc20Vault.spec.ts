import { ethers } from 'hardhat'
import { Contract, Signer, Wallet } from 'ethers'
import { loadFixture } from '@nomicfoundation/hardhat-toolbox/network-helpers'
import {
  EthErc20Vault,
  Keeper,
  OsTokenVaultController,
  VaultsRegistry,
  OsTokenConfig,
  SharedMevEscrow,
  DepositDataRegistry,
} from '../typechain-types'
import snapshotGasCost from './shared/snapshotGasCost'
import {
  createDepositorMock,
  deployEthVaultV1,
  encodeEthErc20VaultInitParams,
  ethVaultFixture,
} from './shared/fixtures'
import { expect } from './shared/expect'
import { ZERO_ADDRESS } from './shared/constants'
import {
  collateralizeEthVault,
  getHarvestParams,
  getRewardsRootProof,
  updateRewards,
} from './shared/rewards'
import keccak256 from 'keccak256'
import { extractExitPositionTicket, setBalance } from './shared/utils'
import { MAINNET_FORK } from '../helpers/constants'
import { registerEthValidator } from './shared/validators'
import { getEthErc20VaultV1Factory } from './shared/contracts'

describe('EthErc20Vault', () => {
  const capacity = ethers.parseEther('1000')
  const name = 'SW ETH Vault'
  const symbol = 'SW-ETH-1'
  const feePercent = 1000
  const metadataIpfsHash = 'bafkreidivzimqfqtoqxkrpge6bjyhlvxqs3rhe73owtmdulaxr5do5in7u'
  const referrer = '0x' + '1'.repeat(40)
  let sender: Wallet, receiver: Wallet, admin: Signer
  let vault: EthErc20Vault,
    keeper: Keeper,
    validatorsRegistry: Contract,
    osTokenVaultController: OsTokenVaultController,
    vaultsRegistry: VaultsRegistry,
    osTokenConfig: OsTokenConfig,
    sharedMevEscrow: SharedMevEscrow,
    depositDataRegistry: DepositDataRegistry

  beforeEach('deploy fixtures', async () => {
    ;[sender, receiver, admin] = (await (ethers as any).getSigners()).slice(1, 4)
    const fixture = await loadFixture(ethVaultFixture)
    vault = await fixture.createEthErc20Vault(admin, {
      capacity,
      name,
      symbol,
      feePercent,
      metadataIpfsHash,
    })
    admin = await ethers.getImpersonatedSigner(await vault.admin())
    keeper = fixture.keeper
    validatorsRegistry = fixture.validatorsRegistry
    osTokenVaultController = fixture.osTokenVaultController
    osTokenConfig = fixture.osTokenConfig
    vaultsRegistry = fixture.vaultsRegistry
    sharedMevEscrow = fixture.sharedMevEscrow
    depositDataRegistry = fixture.depositDataRegistry
  })

  it('has id', async () => {
    expect(await vault.vaultId()).to.eq('0x' + keccak256('EthErc20Vault').toString('hex'))
  })

  it('has version', async () => {
    expect(await vault.version()).to.eq(2)
  })

  it('deposit emits transfer event', async () => {
    const amount = ethers.parseEther('100')
    let expectedShares = await vault.convertToShares(amount)
    if (MAINNET_FORK.enabled) {
      expectedShares += 1n // rounding error
    }

    const receipt = await vault
      .connect(sender)
      .deposit(receiver.address, referrer, { value: amount })
    expect(await vault.balanceOf(receiver.address)).to.eq(expectedShares)

    await expect(receipt)
      .to.emit(vault, 'Deposited')
      .withArgs(sender.address, receiver.address, amount, expectedShares, referrer)
    await expect(receipt)
      .to.emit(vault, 'Transfer')
      .withArgs(ZERO_ADDRESS, receiver.address, expectedShares)
    await snapshotGasCost(receipt)
  })

  it('deposit through receive fallback function emits transfer event', async () => {
    const depositorMock = await createDepositorMock(vault)
    const depositorMockAddress = await depositorMock.getAddress()

    const amount = ethers.parseEther('100')
    let expectedShares = await vault.convertToShares(amount)
    expect(await vault.convertToShares(amount)).to.eq(expectedShares)

    const receipt = await depositorMock.connect(sender).depositToVault({ value: amount })
    if (MAINNET_FORK.enabled) {
      expectedShares += 1n // rounding error
    }
    expect(await vault.balanceOf(depositorMockAddress)).to.eq(expectedShares)

    await expect(receipt)
      .to.emit(vault, 'Deposited')
      .withArgs(depositorMockAddress, depositorMockAddress, amount, expectedShares, ZERO_ADDRESS)
    await expect(receipt)
      .to.emit(vault, 'Transfer')
      .withArgs(ZERO_ADDRESS, depositorMockAddress, expectedShares)
    await snapshotGasCost(receipt)
  })

  it('enter exit queue emits transfer event', async () => {
    await collateralizeEthVault(vault, keeper, depositDataRegistry, admin, validatorsRegistry)
    expect(await vault.totalExitingAssets()).to.be.eq(0)
    const totalExitingBefore = await vault.totalExitingAssets()
    const totalAssetsBefore = await vault.totalAssets()
    const totalSharesBefore = await vault.totalShares()

    const amount = ethers.parseEther('100')
    let shares = await vault.convertToShares(amount)
    await vault.connect(sender).deposit(sender.address, referrer, { value: amount })
    if (MAINNET_FORK.enabled) {
      shares += 1n // rounding error
    }
    expect(await vault.balanceOf(sender.address)).to.be.eq(shares)

    const receipt = await vault.connect(sender).enterExitQueue(shares, receiver.address)
    const positionTicket = await extractExitPositionTicket(receipt)
    await expect(receipt)
      .to.emit(vault, 'V2ExitQueueEntered')
      .withArgs(sender.address, receiver.address, positionTicket, shares, amount)
    await expect(receipt).to.emit(vault, 'Transfer').withArgs(sender.address, ZERO_ADDRESS, shares)
    expect(await vault.totalExitingAssets()).to.be.eq(totalExitingBefore + amount)
    expect(await vault.totalAssets()).to.be.eq(totalAssetsBefore)
    expect(await vault.totalSupply()).to.be.eq(totalSharesBefore)
    expect(await vault.balanceOf(sender.address)).to.be.eq(0)

    await snapshotGasCost(receipt)
  })

  it('update exit queue emits transfer event', async () => {
    const vault = await deployEthVaultV1(
      await getEthErc20VaultV1Factory(),
      admin,
      keeper,
      vaultsRegistry,
      validatorsRegistry,
      osTokenVaultController,
      osTokenConfig,
      sharedMevEscrow,
      encodeEthErc20VaultInitParams({
        capacity,
        name,
        symbol,
        feePercent,
        metadataIpfsHash,
      })
    )
    const validatorDeposit = ethers.parseEther('32')
    await vault
      .connect(admin)
      .deposit(await admin.getAddress(), ZERO_ADDRESS, { value: validatorDeposit })
    await registerEthValidator(vault, keeper, depositDataRegistry, admin, validatorsRegistry)

    const rewardsTree = await updateRewards(keeper, [
      { vault: await vault.getAddress(), reward: 0n, unlockedMevReward: 0n },
    ])
    const proof = getRewardsRootProof(rewardsTree, {
      vault: await vault.getAddress(),
      reward: 0n,
      unlockedMevReward: 0n,
    })

    // exit validator
    await vault.connect(admin).enterExitQueue(validatorDeposit, await admin.getAddress())
    await setBalance(await vault.getAddress(), validatorDeposit)

    const receipt = await vault.updateState({
      rewardsRoot: rewardsTree.root,
      reward: 0n,
      unlockedMevReward: 0n,
      proof,
    })
    await expect(receipt)
      .to.emit(vault, 'Transfer')
      .withArgs(await vault.getAddress(), ZERO_ADDRESS, validatorDeposit)

    await snapshotGasCost(receipt)
  })

  it('cannot transfer vault shares when unharvested and osToken minted', async () => {
    await collateralizeEthVault(vault, keeper, depositDataRegistry, admin, validatorsRegistry)
    const assets = ethers.parseEther('1')
    const shares = await vault.convertToShares(assets)
    const osTokenShares = await osTokenVaultController.convertToShares(assets / 2n)

    await vault.connect(sender).deposit(sender.address, ZERO_ADDRESS, { value: assets })
    await vault.connect(sender).mintOsToken(receiver.address, osTokenShares, ZERO_ADDRESS)

    await updateRewards(keeper, [
      getHarvestParams(await vault.getAddress(), ethers.parseEther('1'), ethers.parseEther('0')),
    ])
    await updateRewards(keeper, [
      getHarvestParams(await vault.getAddress(), ethers.parseEther('1.2'), ethers.parseEther('0')),
    ])
    await expect(
      vault.connect(sender).transfer(receiver.address, shares)
    ).to.be.revertedWithCustomError(vault, 'NotHarvested')
  })

  it('cannot transfer vault shares when LTV is violated', async () => {
    await collateralizeEthVault(vault, keeper, depositDataRegistry, admin, validatorsRegistry)
    const assets = ethers.parseEther('2')
    const shares = await vault.convertToShares(assets)
    const osTokenShares = await osTokenVaultController.convertToShares(assets / 2n)

    await vault.connect(sender).deposit(sender.address, ZERO_ADDRESS, { value: assets })
    await vault.connect(sender).mintOsToken(receiver.address, osTokenShares, ZERO_ADDRESS)
    await expect(
      vault.connect(sender).transfer(receiver.address, shares)
    ).to.be.revertedWithCustomError(vault, 'LowLtv')
    await vault.connect(sender).approve(receiver.address, shares)
    await expect(
      vault.connect(receiver).transferFrom(sender.address, receiver.address, shares)
    ).to.be.revertedWithCustomError(vault, 'LowLtv')
  })

  it('can transfer vault shares when LTV is not violated', async () => {
    await collateralizeEthVault(vault, keeper, depositDataRegistry, admin, validatorsRegistry)
    const assets = ethers.parseEther('2')
    const osTokenShares = await osTokenVaultController.convertToShares(assets / 2n)
    const transferShares = await vault.convertToShares(ethers.parseEther('0.1'))

    await vault.connect(sender).deposit(sender.address, ZERO_ADDRESS, { value: assets })
    await vault.connect(sender).mintOsToken(receiver.address, osTokenShares, ZERO_ADDRESS)
    await expect(vault.connect(sender).transfer(receiver.address, transferShares)).to.emit(
      vault,
      'Transfer'
    )
    await vault.connect(sender).approve(receiver.address, transferShares)
    await expect(
      vault.connect(receiver).transferFrom(sender.address, receiver.address, transferShares)
    ).to.emit(vault, 'Transfer')
  })
})
