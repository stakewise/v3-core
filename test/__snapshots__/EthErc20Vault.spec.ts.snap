// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EthErc20Vault deposit emits transfer event 1`] = `
Object {
  "calldataByteLength": 68,
  "gasUsed": 68858,
}
`;

exports[`EthErc20Vault deposit through receive fallback function emits transfer event 1`] = `
Object {
  "calldataByteLength": 4,
  "gasUsed": 76116,
}
`;

exports[`EthErc20Vault enter exit queue emits transfer event 1`] = `
Object {
  "calldataByteLength": 68,
  "gasUsed": 89043,
}
`;

exports[`EthErc20Vault redeem emits transfer event 1`] = `
Object {
  "calldataByteLength": 68,
  "gasUsed": 60170,
}
`;

exports[`EthErc20Vault update exit queue emits transfer event 1`] = `
Object {
  "calldataByteLength": 196,
  "gasUsed": 137297,
}
`;
